guides/deployment/releases.md

```diff
@@ -150,8 +150,11 @@ Elixir releases work well with container technologies, such as Docker. The idea
 
 Here is an example Docker file to run at the root of your application covering all of the steps above:
 
-```docker
-FROM hexpm/elixir:1.11.2-erlang-23.1.2-alpine-3.12.1 as deps
+```Dockerfile
+FROM hexpm/elixir:1.11.2-erlang-23.1.2-alpine-3.12.1 as build
+
+# install build dependencies
+RUN apk add --no-cache build-base npm git python3 curl
```

Comment:
We can also use the latest update packages while building the release, just like I have recommend for the final release stage:

```
apk upgrade --no-cache
```

Reply:
why do we need `git` dependency here?

Reply:
the same question for curl, you're not using it anywhere

Reply:
It is common to have `git` dependencies in your project mix.exs file. It is not needed in the final build though.

Reply:
Ok, we need `git` in case if we have dependencies in our `mix.exs` which we pull from git instead of hex.pm.

Just got used to include only what project really needs in docker image, I assume we're trying to provide more general image here with many things "just in case" instead of minimal base to grow for each individual project.

Reply:
@achempion `curl` was likely added based on a comment I wrote earlier in the thread:

> Might be worth considering to install curl if you plan to use a Docker HEALTHCHECK which could be defined in a number of spots (Dockerfile or docker-compose.yml).

It's really common to instruct Docker to curl a URL (`http://localhost:8000` or whatever) inside of your container to determine if your service is up or not.

Reply:
@nickjj the final image does now have `curl` with current Dockerfile structure, it's included only on the first step and discard later because multi-stage builds do not preserve anything between stages

Reply:
Anyway, even if a result image had curl installed, not sure how common it is, it seems that [kubernetes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) does not depend of curl presence inside the image.

Reply:
Yeah, if you planned to use Docker's health check feature you'd want it in the final image because Docker will run the health check command inside of the container and curl would need to be available there.

The health check ultimately becomes a command that Docker runs on a set interval (you can define the params of this) and then it'll determine if your app is healthy based on the exit code of the command you're running where 0 is healthy and 1 is unhealthy.

Reply:
> Yeah, if you planned to use Docker's health check feature you'd want it in the final image because Docker will run the health check command inside of the container and curl would need to be available there.

Right, but the final image here still does not have curl installed, it's present only on the first stage of a build where we download/compile dependencies. It's unclear for me why do we need curl on the first stage.

Reply:
> It's unclear for me why do we need curl on the first stage.

Ah ok, that could go either way. Technically it's not needed, but it might be worth adding so you can test your health check in development where you would be using the build stage instead of a release stage.

But truthfully in practice I configure all of my health check commands with env variables and then use `/bin/true` in development to avoid Docker hitting my server every 30 seconds in development which can get a little noisy in the logs if you're trying to debug.

Reply:
@nickjj Please, re-read my comments above, the final image built with current configuration **will not have curl installed**, i.e. **you could not use healthcheck feature** if you build your image with suggested dockerfile.

Reply:
Yeah we went over the final image question in a previous reply:

> Yeah, if you planned to use Docker's health check feature you'd want it in the final image because Docker will run the health check command inside of the container and curl would need to be available there.

You can do whatever you'd like to do in the end. Personally I install curl in both based on my recent replies here.

As for Kubernetes, you're right it doesn't need any special tools but if you plan to use Docker Compose or Swarm you do, unless you decide to include your own separate language specific script which does an HTTP request to a URL and then instruct your health check to call that instead of using curl.

Reply:
I think if there is no HEATHCHECK set in the df, you shouldn't include provisions for it since who ever adds it will know what they want to use and may not use curl.

I would also personally drop git (and python3?, not sure what's depending on that, I haven't needed it in my df's), if someone is using specific git repo/branches in their mix file it becomes pretty clear why the build is failing and they're probably advanced enough to fix it. 

It is in the build stage though so really it comes down to how fastidious you are about having minimal stages.

Reply:
> I think if there is no HEATHCHECK set in the df

Yeah this is true, although you can use a health check without defining `HEALTHCHECK` in your `Dockerfile`. It's much more flexible to define it at runtime since it lets you change it without needing to rebuild your image, and certain container runtimes like Kubernetes ignore it.

Docker Compose / Swarm have support for it in `docker-compose.yml` which configures it at runtime or if you happen to be running `docker` directly it's a flag you can set.

Maybe it's worth calling that out in the docs (as documentation) since having a health check in the end is a really good idea no matter which container runtime strategy you use.

Reply:
I don't think it's a bad idea to include a heathcheck in the dockerfile, just like, include it and curl _or_ include neither. I would even comment on why things are included:

```dockerfile
# install build dependencies
#  - git: for mix git deps
# ...

# install runtime dependencies
#   openssl for
#   ncur...
#   curl for health check 
# note: releases include a built in erlang/elixir don't need to install ...
```

Which does bloat the line count a bit but it helps make the dockerfile a bit less of a black box.

> Maybe it's worth calling that out in the docs (as documentation) since having a health check in the end is a really good idea no matter which container runtime strategy you use.

I think this is a good idea, maybe some links to docker best practices if a reliable, stable source for that exists.